includes/admin/extensions/abstract-extension.php

```diff
@@ -0,0 +1,312 @@
+	 *
+	 * @param array $array
+	 * @return string
+	 */
+	private function get_type( array $array ) {
+		$type = 'product';
+		if ( empty( $array ) ) {
+			return $type;
+		}
+		if ( function_exists( 'array_key_first' ) ) {
+			return array_key_first( $array );
+		}
+		foreach ( $array as $key => $unused ) {
+			return $key;
+		}
+
+		return $type;
+	}
+
+	/**
+	 * Gets the button parameters.
+	 * Classes should not need to replace this method.
+	 *
+	 * @param \EDD\Admin\Extensions\ProductData $product_data The extension data returned from the Products API.
+	 * @param int|false                         $item_id      Optional: the item ID.
+	 * @return array
+	 */
+	protected function get_button_parameters( $product_data, $item_id = false ) {
+		if ( empty( $item_id ) ) {
+			$item_id = $this->item_id;
+		}
+		$body   = $this->get_api_body();
+		$type   = $this->get_type( $body );
+		$id     = ! empty( $body[ $type ] ) ? $body[ $type ] : $this->item_id;
+		$button = array(
+			'type'    => $type,
+			'id'      => $id,
+			'product' => $item_id,
+		);
+		// If the extension is not installed, the button will prompt to install and activate it.
+		if ( ! $this->manager->is_plugin_installed( $product_data->basename ) ) {
+			if ( $this->manager->pass_can_download() ) {
+				$button = array(
+					/* translators: The extension name. */
+					'button_text' => sprintf( __( 'Log In to Your Account to Download %s', 'easy-digital-downloads' ), $product_data->title ),
+					'href'        => $this->get_upgrade_url( $product_data, $item_id, true ),
+					'new_tab'     => true,
+					'type'        => $type,
+				);
+			} else {
+				$button = array(
+					/* translators: The extension name. */
+					'button_text' => sprintf( __( 'Upgrade Today to Access %s!', 'easy-digital-downloads' ), $product_data->title ),
+					'href'        => $this->get_upgrade_url( $product_data, $item_id ),
+					'new_tab'     => true,
+					'type'        => $type,
+				);
+			}
+		} elseif ( ! empty( $product_data->basename ) && ! $this->manager->is_plugin_active( $product_data->basename ) ) {
+			// If the extension is installed, but not activated, the button will prompt to activate it.
+			$button['plugin'] = $product_data->basename;
+			$button['action'] = 'activate';
+			/* translators: The extension name. */
+			$button['button_text'] = sprintf( __( 'Activate %s', 'easy-digital-downloads' ), $product_data->title );
+		}
+
+		return $button;
+	}
+
+	/**
+	 * Gets the upgrade URL for the button.
+	 *
+	 * @todo add UTM parameters
+	 * @since 2.11.x
+	 * @param \EDD\Admin\Extensions\ProductData $product_data The product data object.
+	 * @param int                               $item_id      The item/product ID.
+	 * @param bool                              $has_access   Whether the user already has access to the extension (based on pass level).
+	 * @return string
+	 */
+	private function get_upgrade_url( $product_data, $item_id, $has_access = false ) {
```

Comment:
```suggestion
	private function get_upgrade_url( ProductData $product_data, $item_id, $has_access = false ) {
```

Because we can. :heart: